File: stock-market-service/src/main/java/rs/edu/raf/strategy/AllOrNoneOrderProcessStrategy.java
package rs.edu.raf.strategy;

import com.google.gson.Gson;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import rs.edu.raf.exceptions.StockDoesntExist;
import rs.edu.raf.model.ListingOwner;
import rs.edu.raf.model.Stock;
import rs.edu.raf.model.order.Order;
import rs.edu.raf.model.order.OrderAction;
import rs.edu.raf.repository.ListingOwnerRepository;
import rs.edu.raf.repository.OrderRepository;
import rs.edu.raf.repository.StockRepository;

import java.util.List;

@Component("allOrNoneOrder")
public class AllOrNoneOrderProcessStrategy extends AbstractOrderProcessStrategy {
    private OrderRepository orderRepository;
    private StockRepository stockRepository;
    private ListingOwnerRepository listingOwnerRepository;

    public AllOrNoneOrderProcessStrategy(@Value("${update.balance.exchange}") String UPDATE_BALANCE_EXCHANGE,
                                      @Value("${update.balance.routing.key}") String UPDATE_BALANCE_ROUTING_KEY, OrderRepository orderRepository, ListingOwnerRepository listingOwnerRepository, RabbitTemplate rabbitTemplate, Gson gson, StockRepository stockRepository) {
        super(UPDATE_BALANCE_EXCHANGE, UPDATE_BALANCE_ROUTING_KEY, orderRepository, listingOwnerRepository, rabbitTemplate, gson);
        this.stockRepository = stockRepository;
        this.orderRepository = orderRepository;
        this.listingOwnerRepository = listingOwnerRepository;
    }

    @Override
    public void processOrder(Order order) {
        if(!isOrderExecutable(order)) {
            return;
        }
        String ticker = order.getTicker();

        Stock stock = stockRepository.findById(ticker).orElseThrow(()->new StockDoesntExist(ticker));

        if(order.getOrderAction() == OrderAction.BUY) {
            checkSellOrders(order, stock.getHigh(), stock.getLow());
        }

        else {
            checkBuyOrders(order, stock.getHigh(), stock.getLow());
        }
    }

    @Override
    public boolean isOrderExecutable(Order order) {
        return true;
    }

    private void checkSellOrders(Order order, Double buyPrice, Double sellPrice) {
        List<Order> sellOrders = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
                order.getTicker(), OrderAction.SELL, 0);

        for(Order sellOrder: sellOrders) {
            if(order.getOrderCreator().longValue() == sellOrder.getOrderCreator().longValue())
                continue;

            if(!isOrderExecutable(sellOrder))
                continue;

            ListingOwner orderListingOwner = listingOwnerRepository.findByOwnerAndTicker(order.getOrderCreator(), order.getTicker())
                    .orElse(new ListingOwner(order.getOrderCreator(), order.getTicker(), 0));

            ListingOwner sellOrderListingOwner = listingOwnerRepository.findByOwnerAndTicker(sellOrder.getOrderCreator(), sellOrder.getTicker())
                    .orElseThrow();

            if(sellOrder.getQuantity() >= order.getQuantity()) {
                Integer quantity = order.getQuantity();
                orderListingOwner.setQuantity(orderListingOwner.getQuantity() + quantity);
                sellOrderListingOwner.setQuantity(sellOrderListingOwner.getQuantity() - quantity);
                sellOrder.setQuantity(sellOrder.getQuantity() - quantity);
                order.setQuantity(0);
                orderRepository.save(sellOrder);
                orderRepository.save(order);
                listingOwnerRepository.save(orderListingOwner);
                listingOwnerRepository.save(sellOrderListingOwner);
                updateBankAccountBalance(orderListingOwner.getOwner(), sellOrderListingOwner.getOwner(), buyPrice * quantity,
                        sellPrice * quantity);
                return;
            }
        }
    }

    private void checkBuyOrders(Order order, Double buyPrice, Double sellPrice) {
        List<Order> buyOrders = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
                order.getTicker(), OrderAction.BUY, 0);

        for(Order buyOrder: buyOrders) {
            if(order.getOrderCreator().longValue() == buyOrder.getOrderCreator().longValue())
                continue;

            if(!isOrderExecutable(buyOrder))
                continue;

            ListingOwner buyOrderListingOwner = listingOwnerRepository.findByOwnerAndTicker(buyOrder.getOrderCreator(),
                            buyOrder.getTicker()).orElse(new ListingOwner(buyOrder.getOrderCreator(), buyOrder.getTicker(), 0));

            ListingOwner orderListingOwner = listingOwnerRepository.findByOwnerAndTicker(order.getOrderCreator(), order.getTicker())
                    .orElseThrow();

            if(buyOrder.getQuantity() >= order.getQuantity()) {
                Integer quantity = order.getQuantity();
                orderListingOwner.setQuantity(orderListingOwner.getQuantity() - quantity);
                buyOrderListingOwner.setQuantity(buyOrderListingOwner.getQuantity() + quantity);
                buyOrder.setQuantity(buyOrder.getQuantity() - quantity);
                order.setQuantity(0);
                orderRepository.save(buyOrder);
                orderRepository.save(order);
                listingOwnerRepository.save(orderListingOwner);
                listingOwnerRepository.save(buyOrderListingOwner);
                updateBankAccountBalance(buyOrderListingOwner.getOwner(), orderListingOwner.getOwner(), buyPrice * quantity,
                        sellPrice * quantity);
                return;
            }
        }
    }

}

File: stock-market-service-kotlin/src/main/kotlin/org/example/stockmarketservicekotlin/strategy/AllOrNoneOrderProcessStrategy.kt
package org.example.stockmarketservicekotlin.strategy

import com.google.gson.Gson
import org.example.stockmarketservicekotlin.exceptions.StockDoesntExistException
import org.example.stockmarketservicekotlin.model.ListingOwner
import org.example.stockmarketservicekotlin.model.Stock
import org.example.stockmarketservicekotlin.model.order.Order
import org.example.stockmarketservicekotlin.model.order.OrderAction
import org.example.stockmarketservicekotlin.repository.ListingOwnerRepository
import org.example.stockmarketservicekotlin.repository.OrderRepository
import org.example.stockmarketservicekotlin.repository.StockRepository
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component


@Component("allOrNoneOrder")
class AllOrNoneOrderProcessStrategy(
    @Value("\${update.balance.exchange}") UPDATE_BALANCE_EXCHANGE: String,
    @Value("\${update.balance.routing.key}") UPDATE_BALANCE_ROUTING_KEY: String,
    rabbitTemplate: RabbitTemplate,
    gson: Gson,
    orderRepository: OrderRepository,
    private val stockRepository: StockRepository,
    private val listingOwnerRepository: ListingOwnerRepository,

    ) : AbstractOrderProcessStrategy(
    UPDATE_BALANCE_EXCHANGE, UPDATE_BALANCE_ROUTING_KEY, orderRepository, listingOwnerRepository, rabbitTemplate, gson
) {
    override fun processOrder(order: Order) {
        if (!isOrderExecutable(order)) {
            return
        }
        val ticker: String = order.ticker

        val stock: Stock = stockRepository.findById(ticker).orElseThrow { StockDoesntExistException(ticker) }

        if (order.orderAction == OrderAction.BUY) {
            checkSellOrders(order, stock.high, stock.low)
        } else {
            checkBuyOrders(order, stock.high, stock.low)
        }
    }

    override fun isOrderExecutable(order: Order): Boolean {
        return true
    }

    private fun checkSellOrders(order: Order, buyPrice: Double, sellPrice: Double) {
        val sellOrders: List<Order> = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
            order.ticker, OrderAction.SELL, 0
        )

        for (sellOrder in sellOrders) {
            if (order.orderCreator == sellOrder.orderCreator) continue

            if (!isOrderExecutable(sellOrder)) continue

            val orderListingOwner: ListingOwner =
                listingOwnerRepository.findByOwnerAndTicker(order.orderCreator, order.ticker)
                    .orElse(ListingOwner(null, order.orderCreator, order.ticker, 0))

            val sellOrderListingOwner: ListingOwner =
                listingOwnerRepository.findByOwnerAndTicker(sellOrder.orderCreator, sellOrder.ticker).orElseThrow()

            if (sellOrder.quantity >= order.quantity) {
                val quantity: Int = order.quantity
                orderListingOwner.quantity += quantity
                sellOrderListingOwner.quantity -= quantity
                sellOrder.quantity -= quantity
                order.quantity = 0
                orderRepository.save(sellOrder)
                orderRepository.save(order)
                listingOwnerRepository.save(orderListingOwner)
                listingOwnerRepository.save(sellOrderListingOwner)
                updateBankAccountBalance(
                    orderListingOwner.owner, sellOrderListingOwner.owner, buyPrice * quantity, sellPrice * quantity
                )
                return
            }
        }
    }

    private fun checkBuyOrders(order: Order, buyPrice: Double, sellPrice: Double) {
        val buyOrders: List<Order> = orderRepository.findAllByTickerAndOrderActionAndQuantityGreaterThan(
            order.ticker, OrderAction.BUY, 0
        )

        for (buyOrder in buyOrders) {
            if (order.orderCreator == buyOrder.orderCreator) continue

            if (!isOrderExecutable(buyOrder)) continue

            val buyOrderListingOwner: ListingOwner = listingOwnerRepository.findByOwnerAndTicker(
                buyOrder.orderCreator, buyOrder.ticker
            ).orElse(ListingOwner(null, buyOrder.orderCreator, buyOrder.ticker, 0))

            val orderListingOwner: ListingOwner =
                listingOwnerRepository.findByOwnerAndTicker(order.orderCreator, order.ticker).orElseThrow()

            if (buyOrder.quantity >= order.quantity) {
                val quantity: Int = order.quantity
                orderListingOwner.quantity -= quantity
                buyOrderListingOwner.quantity += quantity
                buyOrder.quantity -= quantity
                order.quantity = 0
                orderRepository.save(buyOrder)
                orderRepository.save(order)
                listingOwnerRepository.save(orderListingOwner)
                listingOwnerRepository.save(buyOrderListingOwner)
                updateBankAccountBalance(
                    buyOrderListingOwner.owner, orderListingOwner.owner, buyPrice * quantity, sellPrice * quantity
                )
                return
            }
        }
    }
}